src/lib.rs

```diff
@@ -194,15 +195,20 @@ impl Client {
     async fn access_token(&self) -> Result<MutexGuard<'_, AccessToken>, Error> {
         let mut cache = self.cache.lock().await;
         if Instant::now() >= cache.expires {
-            *cache = self
-                .token_request(&match &cache.refresh_token {
-                    Some(refresh_token) => (
-                        ("grant_type", "refresh_token"),
-                        Some(("refresh_token", refresh_token)),
-                    ),
-                    None => (("grant_type", "client_credentials"), None),
-                })
-                .await?;
+            *cache = match &cache.refresh_token {
+                // Authorization code flow
+                Some(refresh_token) => {
+                    self.token_request(TokenRequest::RefreshToken { refresh_token })
+                        .map_ok(|mut token| {
+                            // Keep hold of refresh_token for next round
+                            token.refresh_token = Some(refresh_token.clone());
+                            token
+                        })
+                        .await?
```

Comment:
Moving this `.await?` to the call on `token_request` avoids having to use `map_ok`; you can just use normal control flow.